<h1>Loginom 6.4.3 — Release Notes</h1>

<p>In this update, file storage was finalized, performance improvements related to operation of processor and memory in the multithreading mode were made, and some fixes were performed.</p>

<h2 id="исправление-ошибок">Fixes</h2>

<p><a href="https://help.loginom.ru/adminguide/server" target="_blank"><strong>Loginom Server:</strong></a>&nbsp;when several sessions with interconnected packages were completed simultaneously, the following error could occur: "Index value of Items parameter is out of range" that was recorded in the server log (#8470).</p>

<p><strong>Loginom Server:</strong>&nbsp;when using Apache web server with the configuration created by default, connection via Web Socket Proxy failed. The error was caused by the fact that&nbsp;<em>WS_PROXY_URI</em>&nbsp;parameter in the created &nbsp;<em>env.conf</em> file was set as&nbsp;<em>ws/</em>. The correct paramater value is&nbsp;<em>app/ws/</em>&nbsp;(#8544).</p>

<p><a href="https://help.loginom.ru/userguide/processors/integration/soap-request.html" target="_blank"><strong>SOAP request:</strong></a>&nbsp;if WSDL schema contained two items with the same names one of which was the root one, the following error occurred while request sending: "Input XSD schema does not comply with the node settings!" (#6179).</p>

<p><a href="https://help.loginom.ru/userguide/visualization/preview/quick-view.html" target="_blank"><strong>Quick view:</strong></a>&nbsp;in the Quick view detailing, values of the items with "Variable" type were displayed as&nbsp;(#8630).</p>

<p><a href="https://help.loginom.ru/userguide/processors/control/execute-node.html" target="_blank"><strong>Node Execution:</strong></a>&nbsp;if in "Node Execution" settings the node for which the execution order had been earlier set as the selected one in such a way that "Node Execution" node was to be performed earlier than the selected node, the following error occurred: "Next move forbidden. Circular links between nodes are not allowed". The same error could occur while opening the package containing the same node configuration. Such situation does not represent a circular reference as only node configuration is used for "Node Execution" as opposed to "Reference Node", and data that differ from the data that is used in the base node can be supplied to the input (#8614).</p>

<p><a href="https://help.loginom.ru/userguide/processors/transformation/trans-datatime/" target="_blank"><strong>Date and Time:</strong></a>&nbsp;when transforming dates to the string format, there were no leading zeros for year week, quarter week, year day and quarter day. The default month day format was with leading zero. The default Date and Time format was Fixed (<em>"2021-01-30"</em>,&nbsp;<em>"23:59:59"</em>,&nbsp;<em>"01:01:01"</em>) (#8595).</p>

<p><a href="https://help.loginom.ru/userguide/integration/import/warehouse.html" target="_blank"><strong>Import from Deductor Warehouse:</strong></a>&nbsp;if the package with configured import from Deductor Warehouse is saved and repeatedly opened, while attempting to execute the import node, the following error occurred: "Deductor Warehouse: Data Warehouse object not specified" (#8523).</p>

<p><strong>Import from Deductor Warehouse:</strong>&nbsp;if one and the same Dimension was included into the Process several times, and the caption was changed while designing, these Dimensions had the same names in the import wizard in the parameters tree, when transferring such Dimensions to the Fields area, their caption texts were incorrectly displayed (#8556).</p>

<p><a href="https://help.loginom.ru/userguide/integration/connections/list/postgresql.html" target="_blank"><strong>Import from PostgreSQL:</strong></a>&nbsp;when one and the same parameter was used more than once in the request, the following error occurred: "Invalid parameters specified" (#8587).</p>

<p><a href="https://help.loginom.ru/userguide/visualization/binary-classification/" target="_blank"><strong>Binary classification assessment:</strong></a>&nbsp;when copying data from error matrix to the clipboard to the values in&nbsp;<em>Predicted</em> column&nbsp;odd&nbsp;<em>%</em> symbols were added (#8638).</p>

<p><a href="https://help.loginom.ru/userguide/processors/preprocessing/fine-classes.html" target="_blank"><strong>Coarse сlasses:</strong></a>&nbsp;when moving in the wizard back from the binning configuration page and then backwards, if the settings that had influence on class count had been changed, the previous class count was displayed on the wizard pages. It could also cause the following errors: "Next move forbidden Failed to apply fine classes for binning for %s column. There are too many unique values" or "Value 0 of "Index" parameter is out of [-1; -1] range" when switching the wizard pages (#8282, #8582, #8584, #8668, #8700).</p>

<p><strong>Coarse сlasses:</strong> while merging classes with continuous data type, the upper and lower bounds of the updated classes were incorrectly defined (#8577).</p>

<p><strong>Coarse сlasses:</strong> "Merge with previous" and "Merge with next" buttons were always active in the binning editing mode. If one of the buttons was pressed upon classes merging, classes were merged, in spite of the fact that none of the classes had been selected. Now, when enabling the editing mode, the selected class is clearly displayed, and "Merge with previous" and "Merge with next" buttons become inactive if the action cannot be performed (#8578).</p>

<p><a href="https://help.loginom.ru/userguide/visualization/cube/" target="_blank"><strong>Cube:</strong></a>&nbsp;if after import from Deductor Warehouse or LGD file in which <em>Group</em> default usage types were set for fields, the Cube visualizer was created, and the displayed measures were configured&nbsp;for the fields with <em>Group</em> default usage types. Then, the calculated measure based on the measures selected earlier for display was added. Afterwards, the measures used for calculation of the calcualted measures were concealed, and after package saving and repeated opening, the calculated measures were not calculated (#5803).</p>

<p><strong>Cube:</strong>&nbsp;while package saving, order of the fields transferred to the Cube filtering area was not saved (#8490).</p>

<p><strong>Cube:</strong>&nbsp;if the number of dimensions in columns was higher than the number of columns in the cross table, some part of dimension captions was not displayed when exporting. Such situation could arise if column totals display was disabled (#8639).</p>

<p><strong>Cube, <a href="https://help.loginom.ru/userguide/processors/control/submodel.html" target="_blank">Supernode:</a></strong>&nbsp;if the Cube visualizer with the sorting configured by one of the columns was configured in the base Supernode, while attempting to configure the derived Supernode, the following error occurred: "__baseName attribute does not exist" (#8536).</p>

<p><strong>Cube, Supernode:</strong>&nbsp;if sorting was configured for the node containing the configured Cube visualizer in the derived Supernode, and then dimensions in the Cube were added for the base Supernode node, the configured visualizer was marked as the invalid one upon the node configuration update (#8528).</p>

<p><a href="https://help.loginom.ru/userguide/scenario/batchlauncher.html" target="_blank"><strong>Batch processing:</strong></a>&nbsp;if the package file to which the processed package referred was deleted or renamed during the package batch processing, upon the main package processing, the following message was recorded to the server log: "Error accessing file: ...lgp", and the processed package remained in the server memory, whereas the deleted or renamed package was still displayed in the list of common packages in the Session Manager  (#8572).</p>

<p><strong><a href="https://help.loginom.ru/userguide/integration/connections/list/firebird.html" target="_blank">Connections to FireBird DB,</a> <a href="https://help.loginom.ru/userguide/integration/connections/list/interbase.html" target="_blank">Interbase:</a></strong>&nbsp;upon database server connection loss, connection was not restored on repeated connection activation (#8615).</p>

<p><strong>Supernode:</strong>&nbsp;if the node connected to the output Supernode port was activated, and the Supernode itself was not activated, and then the package containing the Supernode was saved and closed, on repeated package opening and while attempting to activate the Supernode, the following error occurred: "Port not configured" (#8538).</p>

<p><strong>Supernode:</strong>&nbsp;if the base Supernode was changed upon creation of the derived component and derived node (for example, one more node was added to the supernode), while attempting to activate the derived node, the following error occurred: "Internal Supernode nodes locked". The error occurred in version 6.4.2 (#8629).</p>

<p><a href="https://help.loginom.ru/userguide/visualization/table/" target="_blank"><strong>Table:</strong></a>&nbsp;the value of <em>Base interval date</em> parameter was not defined and saved in the filter configuration window in the Table visualizer for the fields of&nbsp;the <em>Date/Time</em> type,&nbsp;when selecting&nbsp;<em>first / not first</em>&nbsp;or&nbsp;<em>last / not last</em>&nbsp;conditions, &nbsp;(#8648).</p>

<p><a href="https://help.loginom.ru/userguide/processors/data-trees/table-to-tree.html" target="_blank"><strong>Table to tree:</strong></a>&nbsp;if the automatic mapping of the input and output fields failed, the following error occurred upon the node configuration: "Table to tree: The output node ... has not been mapped to the input node". It was required to go to the output port settings to perform the fixes. Now, if such situation arises during the node configuration process, the following page is opened: "Adjustment of Data Trees Mapping" (#8659).</p>

<p><a href="https://help.loginom.ru/userguide/processors/control/unit-link.html" target="_blank"><strong>Reference node:</strong></a>&nbsp;when saving the package containing the Reference Node, the following error could occur: "..." node is locked" if the node that used the data of the output port of the source node for which the Reference Node had been configured (#8680) was executed in the parallel session.</p>

<p><a href="https://help.loginom.ru/userguide/location_user_files.html" target="_blank"><strong>File storage:</strong></a>&nbsp;if the new directories became available to a user during his current session, while attempting to go to such directory, the following error occurred: "Cannot read property 'expandAsync' of undefined". For example, such situation could arise if another user gave access to the user, the rights of access to the directories via the Administration bar were changed, or the new directory was created via the file saving dialog (#5347, #6032, #6423, #8022).</p>

<p><strong>File storage:</strong>&nbsp;if the internal directory was moved one level up or down from the directory, while attempting to open the first directory, the following error occurred: "System Error. Code: 3. System failed to find the specified path" (#6674).</p>

<p><strong>File storage:</strong>&nbsp;if one of the directores containing the copied directory was renamed after directory copying, when pasting, the following error occurred: "System Error. Code: 3. System failed to find the specified path" (#7082).</p>

<p><strong>File storage:</strong>&nbsp;if the file was downloaded and then deleted, it was impossible to download it again with the same name before the application session termination. The error occurred in the browsers based on Chromium, in Internet Explorer and Safari (#7108, #8058, #8499).</p>

<p><strong>File storage:</strong>&nbsp;when moving from the current directory to the parent one,&nbsp;<em>".."</em> was selected but not the quitted directory. Из-за неправильной установки фокуса мог некорректно работать функционал копирования–вставки через горячие клавиши (#8023, #8516).</p>

<p><strong>Файловое хранилище:</strong>&nbsp;при попытке открыть пакет через символическую ссылку на файл, физически расположенный в сетевой папке, из-за невозможности определить дату модификации файла, возникала ошибка «Системе не удается найти указанный путь» (#8452).</p>

<p><strong>File storage:</strong>&nbsp;the folder deleted by means of "Cut-Paste" option was not deleted from the directory tree. While attempting to move to it in the directory tree, the following error occurred: "System Error". Code: 3. System failed to find the specified path" (#8518).</p>

<p><strong>Файловое хранилище:</strong>&nbsp;если&nbsp;перейти в папку файлового хранилища, ранее открытую в другой вкладке, а затем попытаться закрыть вкладку с ранее открытой папкой нажатием на&nbsp;<em>X</em>&nbsp;(крестик), то возникала ошибка "Cannot read property 'ActiveLoadersCount' of undefined" или "this.FFileSenderManager is undefined" (#8529).</p>

<p><a href="https://help.loginom.ru/userguide/processors/transformation/row-filter/" target="_blank"><strong>Фильтр строк:</strong></a>&nbsp;при повторном заходе в мастер фильтрации метка на плитке с фильтром по временному промежутку отображалась не полностью, если текст был длинный. Например, если было выбрано условие «первый день от текущей даты» (#8633).</p>

<p><strong>Фильтр строк:</strong>&nbsp;было невозможно настроить два и более условий по временному промежутку (<em>первый/не первый или последний/не последний</em>) (#8634).</p>

<p><a href="https://help.loginom.ru/userguide/processors/control/cycle.html" target="_blank"><strong>Цикл:</strong></a>&nbsp;при активации параллельного Цикла, содержащего вложенную Подмодель и Узел-ссылку, ссылающийся на узел, находящийся внутри Подмодели, могла возникать ошибка «Узел "Цикл" заблокирован». Чем больше итераций было выбрано в настройках Цикла, тем выше была вероятность возникновения ошибки (#8531).</p>

<p><strong>Export to PostgreSQL:</strong>&nbsp;while attempting to export strings to&nbsp;<em>xml</em>&nbsp;or&nbsp;<em>json</em> fields,&nbsp;the following error occurred: "column "..." has type ..., and expression - character varying". Если ранее возникала ошибка, то после обновления необходимо повторно пройти мастер настройки экспорта (#8585).</p>

<p><strong>Ядро платформы:</strong>&nbsp;в случае, если Подмодель содержала Узел-ссылку, ссылающийся на другой узел внутри этой подмодели, а порядок выполнения для Узла-ссылки был связан со входным и выходным портом содержащей его Подмодели, то процесс активации Подмодели мог зависнуть и его невозможно было отменить. If the Supernode was executed in the Loop, it increased the risk of error occurrence(#8514).</p>

<h2 id="повышение-информативности">UI/UX Improvements</h2>

<p><strong>Import from Deductor Warehouse:</strong>&nbsp;while attempting to execute unconfigured import node, uninformative error occurred. Теперь появляется предупреждение, что не выбран объект хранилища данных (#8515).</p>

<p><a href="https://help.loginom.ru/userguide/scenario/ports/" target="_blank"><strong>Настройка портов:</strong></a>&nbsp;сообщения об ошибках «Сокет не настроен» и «Не задана ссылка на сокет» заменены на более информативные сообщения «...: Порт ... не настроен» и «...: К обязательному порту ... не подключена связь» с указанием метки порта, в котором возникла ошибка (#6000).</p>

<h2 id="удобство-использования">Ease of Use</h2>

<p><a href="https://help.loginom.ru/userguide/processors/programming/python/" target="_blank"><strong>Python:</strong></a>&nbsp;функция&nbsp;<em>builtin_pandas_utils.to_data_frame</em>&nbsp;доработана таким образом, чтобы была возможность создания&nbsp;<em>pandas.DataFrame</em>&nbsp;по целочисленным и логическим столбцам, содержащим пропуски. Если при заполнении массивов&nbsp;<em>numpy</em>&nbsp;возникает необходимость обработки пропущенных значений целочисленного и логического типов, то используется вещественный тип, а пропущенные значения преобразуются в&nbsp;<em>NaN</em>&nbsp;(<em>np.nan</em>). Ранее при попытке передать в функцию&nbsp;<em>builtin_pandas_utils.to_data_frame</em>&nbsp;таблицу, содержащую целочисленный или логический столбец с пропущенными значениями, возникала ошибка "...int() argument must be a string, a bytes-like object or a number, not 'NoneType' " (#8522).</p>

<p><strong>Импорт из Deductor Warehouse:</strong>&nbsp;у выходных столбцов отключена автоматическая инициализация «Назначения по умолчанию» на основе типов объектов хранилища данных. Ранее для&nbsp;<em>Фактов</em>&nbsp;устанавливались «Назначения по умолчанию» равное&nbsp;<em>Показатель</em>, а для&nbsp;<em>Измерений</em>&nbsp;-&nbsp;<em>Группа</em>. У Пользователя есть возможность самостоятельно установить для выходных столбцов «Назначения по умолчанию». Прежнее поведение могло, в том числе, приводить к ошибкам в визуализаторе Куб&nbsp;<em>#5803</em>&nbsp;(#8506).</p>

<p><strong>Конечные классы:</strong>&nbsp;отключено отображение с округлением значений границ конечного класса в мастере настройки. Ранее при ручном редактировании границ, их значение отображалось с округлением до двух знаков после запятой (#8635).</p>

<p><strong>Логирование:</strong>&nbsp;в сообщения лога для событий начала и окончания выполнения узла добавлен GUID (уникальный идентификатор) узла (#8667).</p>

<p><strong>Файловое хранилище:</strong>&nbsp;при выполнении функционала «Вырезать-Вставить» вместо копирования и удаления исходной папки теперь используется операция перемещения. Ранее операция «Вырезать-Вставить» могла занимать продолжительное время, а также при копировании каталогов с объемом больше 2 Гб можно было в процессе копирования на соседней вкладке вырезать подкаталог копируемого каталога (#6031, #6366).</p>

<p><strong>Файловое хранилище:</strong>&nbsp;оптимизирована скорость создания каталогов для случая, когда в текущем каталоге уже содержится большое количество каталогов (несколько десятков или сотни вложенных каталогов) (#8196).</p>

<p><strong>Файловое хранилище:</strong>&nbsp;в корне файлового хранилища сделаны неактивными действия, которые невозможно было выполнить:&nbsp;<em>Загрузить, Создать каталог, Скачать, Переименовать, Копировать, Вставить, Вырезать, Удалить</em>&nbsp;(#8542).</p>

<p><strong>Ядро платформы:</strong>&nbsp;количество рабочих потоков сделано динамическим. Если процессор загружен меньше, чем на 80% и есть очередь процессов на выполнение, то количество рабочих потоков увеличивается, пока загрузка процессора не превысит 80%. Если дополнительные потоки простаивают, то они постепенно удаляются, чтобы снизить потребление оперативной памяти. Ранее, если на сервере было одновременно запущено большое количество сценариев, содержащих узлы, которые ожидают ответы от внешних приложений или сервисов (Выполнение программы, REST-запрос, импорт из Баз данных и тому подобные), то могла наблюдаться ситуация, когда процессор загружен значительно меньше 100%, но новые процессы не выполняются (#7255).</p>
